dashboard/test/models/contact_rollups_pardot_memory_test.rb

```diff
@@ -23,61 +23,177 @@ class ContactRollupsPardotMemoryTest < ActiveSupport::TestCase
+      {email: 'alpha', data: {opt_in: false}, data_updated_at: base_time},
+      # Case 2: data_updated_at > data_synced_at
+      {email: 'beta', data: {opt_in: false}, data_updated_at: base_time},
+      # Case 3: pardot_id_updated_at > data_synced_at
+      {email: 'gamma', data: {opt_in: true}, data_updated_at: base_time},
+      # dummy records
+      {email: 'delta'},
+      {email: 'zeta'},
+    ]
+    processed_contact_records.each {|record| create :contact_rollups_processed, record}
+
+    # Execute SQL query
+    results = ActiveRecord::Base.connection.
+      exec_query(ContactRollupsPardotMemory.query_updated_contacts).map do |record|
+      record.slice('email', 'pardot_id_changed')
     end
+
+    expected_results = [
+      {'email' => 'alpha', 'pardot_id_changed' => nil},
+      {'email' => 'beta', 'pardot_id_changed' => 0},
+      {'email' => 'gamma', 'pardot_id_changed' => 1}
+    ]
+    assert_equal expected_results, results
   end
 
-  test 'save_sync_results' do
+  test 'update_pardot_prospects' do
+    email = 'test@domain.com'
+    last_sync_time = Time.now.utc - 7.days
+    create :contact_rollups_pardot_memory, email: email, data_synced: {db_Opt_In: 'No'}, data_synced_at: last_sync_time
+    create :contact_rollups_processed, email: email, data: {'opt_in' => true}
+
+    PardotV2.expects(:submit_batch_request).once.returns([])
+
+    ContactRollupsPardotMemory.update_pardot_prospects
+
+    record = ContactRollupsPardotMemory.find_by(email: email)
+    assert_equal({'db_Opt_In' => 'Yes'}, record.data_synced)
+    assert last_sync_time < record.data_synced_at
+  end
+
+  test 'save_sync_results new prospect' do
     assert_equal 0, ContactRollupsPardotMemory.count
 
-    submissions = [
-      {email: 'invalid_email', id: nil, db_Opt_in: 'No'},
-      {email: 'valid@domain.com', id: nil, db_Opt_in: 'Yes'},
+    submission = {email: 'valid@domain.com', db_Opt_In: 'Yes'}
+    submitted_time = Time.now.utc
+
+    ContactRollupsPardotMemory.save_sync_results [submission], [], submitted_time
+
+    record = ContactRollupsPardotMemory.find_by(
+      email: submission[:email],
+      data_synced_at: submitted_time
+    )
+    expected_data_synced = submission.except(:email, :id).deep_stringify_keys
+    assert_equal expected_data_synced, record.data_synced
+  end
+
+  test 'save_sync_results updated prospects' do
+    assert_equal 0, ContactRollupsPardotMemory.count
+    pardot_memory_records = [
+      {email: 'alpha', pardot_id: 1, data_synced: nil},
+      {email: 'beta', pardot_id: 2, data_synced: {db_Opt_In: 'No'}},
+      {email: 'gamma', pardot_id: 3, data_synced: {db_Opt_In: 'Yes'}},
     ]
-    errors = [
-      {prospect_index: 0, error_msg: PardotHelpers::ERROR_INVALID_EMAIL}
+    pardot_memory_records.each {|record| create :contact_rollups_pardot_memory, record}
+
+    submissions = [
+      {email: 'alpha', id: 1, db_Opt_In: 'Yes'},
+      {email: 'beta', id: 2, db_Opt_In: 'Yes'},
```

Comment:
why is this db_Opt_In yes but the above beta db_Opt_In is no?

Reply:
Because data in `pardot_memory_records` are existing data and should be replaced by newer data in `submissions`. The assertions at the end of this test are meant to verify that.